snapcraft/plugins/copy.py

```diff
@@ -27,23 +29,41 @@ class CopyPlugin(snapcraft.BasePlugin):
 
     @classmethod
     def schema(cls):
-        return {
-            'properties': {
-                'files': {
-                    'type': 'object',
-                },
-            },
-            'required': [
-                'files',
-            ]
+        schema = super().schema()
+        schema['properties']['files'] = {
+            'type': 'object',
         }
+        if 'required' not in schema:
+            schema['required'] = []
+        schema['required'].append('files')
+        return schema
 
     def build(self):
-        for src in sorted(self.options.files):
-            dst = self.options.files[src]
-            if not os.path.lexists(src):
-                raise EnvironmentError('file "{}" missing'.format(src))
+        # Sources handling first
+        super().build()
+
+        for srcname in sorted(self.options.files):
+            dst = self.options.files[srcname]
+
+            # Expand directories to include part info
             dst = os.path.join(self.installdir, dst)
-            os.makedirs(os.path.dirname(dst), exist_ok=True)
-            self.run(['cp', '--preserve=all', '-R', src, dst],
-                     cwd=os.getcwd())
+            src = os.path.join(self.sourcedir, srcname)
+
+            if not os.path.lexists(src):
+                # If it doesn't exist, check if it is a glob
+                filelst = glob.glob(src)
+                if True not in map(os.path.lexists, filelst):
```

Comment:
this feels a bit involved, I'm trying to think of a better way to get this going.